File: music.ts
import { schedule } from "./schedule.ts";
import { Webhook } from "./webhook.ts";
import config from "./config.ts";

const webhook = new Webhook(config.musicWebhook);
const listentothisEndpoint =
  "https://www.reddit.com/r/listentothis/top.json?t=week&limit=1";

Deno.permissions.request({
  name: "net",
  host: "www.reddit.com",
});

async function getMusic() {
  // Get the top song from r/listentothis
  const response = await fetch(listentothisEndpoint, {
    headers: {
      "User-Agent": "Discord Bot",
      Accept: "application/json",
    },
  });
  const data = await response.json();
  const { url } = data.data.children[0].data;
  return url;
}

schedule({
  description:
    "Post the weekly top r/listentothis song to Discord every Saturday at 8AM Pacific Time",
  time: "Saturday 08:00:00",
  async callback() {
    const url = await getMusic();
    webhook.post(url);
  },
});

File: events/music.py
import logging
import os

import requests
from schedule import every, repeat

webhook_url = os.environ["MUSIC_WEBHOOK"]

logger = logging.getLogger("schedule.music")


@repeat(every().saturday.at("08:00"))
def post_music():
    """Post this week's top r/listentothis song to Discord."""

    logger.info("Posting top song")

    # Get the top song from r/listentothis
    try:
        response = requests.get(
            "https://www.reddit.com/r/listentothis/top.json?t=week&limit=1",
            headers={"User-Agent": "Discord Bot"},
        )
        response.raise_for_status()
        data = response.json()
    except requests.HTTPError as exc:
        logger.error(f"Failed to get top song: {exc}")
        return None
    except requests.JSONDecodeError as exc:
        logger.error(f"Failed to decode top song response: {exc}")
        return None

    music_url = data.get("data", {}).get("children", [{}])[0].get("data", {}).get("url")

    if music_url is None:
        logger.error("Failed to get top song")
        return

    try:
        response = requests.post(
            webhook_url,
            json={"content": music_url, "username": "Song-A-Week"},
        )
        response.raise_for_status()

    except requests.HTTPError as exc:
        logging.error(f"Failed to post top song: {exc}")


logger.info("Loaded Song-A-Week event")
